// @filestack/angular/esm5/lib/picker/pickerDropPane.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ChangeDetectionStrategy } from '@angular/core';
import { PickerBaseComponent } from './pickerBase.component';
import { PickerDisplayMode } from 'filestack-js';
var PickerDropPaneComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PickerDropPaneComponent, _super);
    function PickerDropPaneComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    PickerDropPaneComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _a = this, filestackService = _a.filestackService, uploadSuccess = _a.uploadSuccess, uploadError = _a.uploadError, pickerOptions = _a.pickerOptions;
        // Overwrite display mode to be always 'dropPane' in this component
        this.picker = filestackService.picker(tslib_1.__assign({}, pickerOptions, { displayMode: PickerDisplayMode.dropPane, container: 'ng-picker-drop-pane-container', onUploadDone: (/**
             * @param {?} res
             * @return {?}
             */
            function (res) { return uploadSuccess.next(res); }) }));
        // Picker open success handler there is ommited, because it's accessible from pickerOptions
        this.picker.open()
            .then(void (0))
            .catch((/**
         * @param {?} err
         * @return {?}
         */
        function (err) { return uploadError.next(err); }));
    };
    PickerDropPaneComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-picker-drop-pane',
                    template: '<div id="ng-picker-drop-pane-container" class="ng-picker"></div>',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    return PickerDropPaneComponent;
}(PickerBaseComponent));
export { PickerDropPaneComponent };
