// @filestack/angular/esm5/lib/picker/pickerOverlay.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ChangeDetectionStrategy, HostListener, ViewChild, ElementRef } from '@angular/core';
import { PickerBaseComponent } from './pickerBase.component';
import { PickerDisplayMode } from 'filestack-js';
var PickerOverlayComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PickerOverlayComponent, _super);
    function PickerOverlayComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    PickerOverlayComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _a = this, filestackService = _a.filestackService, uploadSuccess = _a.uploadSuccess, pickerOptions = _a.pickerOptions;
        // Overwrite display mode to be always 'overlay' in this component
        this.picker = filestackService.picker(tslib_1.__assign({}, pickerOptions, { displayMode: PickerDisplayMode.overlay, onUploadDone: (/**
             * @param {?} res
             * @return {?}
             */
            function (res) { return uploadSuccess.next(res); }) }));
    };
    /**
     * @return {?}
     */
    PickerOverlayComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _a = this, picker = _a.picker, customRenderRef = _a.customRenderRef, uploadError = _a.uploadError;
        if (!customRenderRef.nativeElement.children.length) {
            // Picker open success handler there is ommited, because it's accessible from pickerOptions
            picker.open()
                .then(void (0))
                .catch((/**
             * @param {?} err
             * @return {?}
             */
            function (err) { return uploadError.next(err); }));
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PickerOverlayComponent.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _a = this, picker = _a.picker, uploadError = _a.uploadError;
        event.stopPropagation();
        event.preventDefault();
        // Picker open success handler there is ommited, because it's accessible from pickerOptions
        picker.open()
            .then(void (0))
            .catch((/**
         * @param {?} err
         * @return {?}
         */
        function (err) { return uploadError.next(err); }));
    };
    PickerOverlayComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-picker-overlay',
                    template: '<div #customRender><ng-content class="ng-picker"></ng-content></div>',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PickerOverlayComponent.propDecorators = {
        customRenderRef: [{ type: ViewChild, args: ['customRender', { static: false },] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return PickerOverlayComponent;
}(PickerBaseComponent));
export { PickerOverlayComponent };
if (false) {
    /** @type {?} */
    PickerOverlayComponent.prototype.customRenderRef;
}
