// @filestack/angular/esm5/lib/picker/pickerBase.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Output, Inject } from '@angular/core';
import { Subject } from 'rxjs';
import { FilestackService } from '../filestack.service';
/**
 * @abstract
 */
var PickerBaseComponent = /** @class */ (function () {
    function PickerBaseComponent(filestackService) {
        this.filestackService = filestackService;
        this.uploadSuccess = new Subject();
        this.uploadError = new Subject();
    }
    /**
     * @return {?}
     */
    PickerBaseComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _a = this, apikey = _a.apikey, clientOptions = _a.clientOptions, filestackService = _a.filestackService;
        filestackService.init(apikey, clientOptions);
    };
    /**
     * @return {?}
     */
    PickerBaseComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        var picker = this.picker;
        if (picker) {
            picker.close();
        }
    };
    /** @nocollapse */
    PickerBaseComponent.ctorParameters = function () { return [
        { type: FilestackService, decorators: [{ type: Inject, args: [FilestackService,] }] }
    ]; };
    PickerBaseComponent.propDecorators = {
        apikey: [{ type: Input }],
        pickerOptions: [{ type: Input }],
        clientOptions: [{ type: Input }],
        file: [{ type: Input }],
        source: [{ type: Input }],
        uploadSuccess: [{ type: Output }],
        uploadError: [{ type: Output }]
    };
    return PickerBaseComponent;
}());
export { PickerBaseComponent };
if (false) {
    /** @type {?} */
    PickerBaseComponent.prototype.apikey;
    /** @type {?} */
    PickerBaseComponent.prototype.pickerOptions;
    /** @type {?} */
    PickerBaseComponent.prototype.clientOptions;
    /** @type {?} */
    PickerBaseComponent.prototype.file;
    /** @type {?} */
    PickerBaseComponent.prototype.source;
    /** @type {?} */
    PickerBaseComponent.prototype.uploadSuccess;
    /** @type {?} */
    PickerBaseComponent.prototype.uploadError;
    /** @type {?} */
    PickerBaseComponent.prototype.picker;
    /**
     * @type {?}
     * @protected
     */
    PickerBaseComponent.prototype.filestackService;
}
