// @filestack/angular/esm2015/lib/picker/pickerDropPane.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy } from '@angular/core';
import { PickerBaseComponent } from './pickerBase.component';
import { PickerDisplayMode } from 'filestack-js';
export class PickerDropPaneComponent extends PickerBaseComponent {
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        const { filestackService, uploadSuccess, uploadError, pickerOptions } = this;
        // Overwrite display mode to be always 'dropPane' in this component
        this.picker = filestackService.picker(Object.assign({}, pickerOptions, { displayMode: PickerDisplayMode.dropPane, container: 'ng-picker-drop-pane-container', onUploadDone: (/**
             * @param {?} res
             * @return {?}
             */
            res => uploadSuccess.next(res)) }));
        // Picker open success handler there is ommited, because it's accessible from pickerOptions
        this.picker.open()
            .then(void (0))
            .catch((/**
         * @param {?} err
         * @return {?}
         */
        err => uploadError.next(err)));
    }
}
PickerDropPaneComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-picker-drop-pane',
                template: '<div id="ng-picker-drop-pane-container" class="ng-picker"></div>',
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
