// @filestack/angular/esm2015/lib/picker/pickerBase.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Output, Inject } from '@angular/core';
import { Subject } from 'rxjs';
import { FilestackService } from '../filestack.service';
/**
 * @abstract
 */
export class PickerBaseComponent {
    /**
     * @param {?} filestackService
     */
    constructor(filestackService) {
        this.filestackService = filestackService;
        this.uploadSuccess = new Subject();
        this.uploadError = new Subject();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        const { apikey, clientOptions, filestackService } = this;
        filestackService.init(apikey, clientOptions);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        const { picker } = this;
        if (picker) {
            picker.close();
        }
    }
}
/** @nocollapse */
PickerBaseComponent.ctorParameters = () => [
    { type: FilestackService, decorators: [{ type: Inject, args: [FilestackService,] }] }
];
PickerBaseComponent.propDecorators = {
    apikey: [{ type: Input }],
    pickerOptions: [{ type: Input }],
    clientOptions: [{ type: Input }],
    file: [{ type: Input }],
    source: [{ type: Input }],
    uploadSuccess: [{ type: Output }],
    uploadError: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PickerBaseComponent.prototype.apikey;
    /** @type {?} */
    PickerBaseComponent.prototype.pickerOptions;
    /** @type {?} */
    PickerBaseComponent.prototype.clientOptions;
    /** @type {?} */
    PickerBaseComponent.prototype.file;
    /** @type {?} */
    PickerBaseComponent.prototype.source;
    /** @type {?} */
    PickerBaseComponent.prototype.uploadSuccess;
    /** @type {?} */
    PickerBaseComponent.prototype.uploadError;
    /** @type {?} */
    PickerBaseComponent.prototype.picker;
    /**
     * @type {?}
     * @protected
     */
    PickerBaseComponent.prototype.filestackService;
}
