// @filestack/angular/esm2015/lib/picker/pickerOverlay.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, HostListener, ViewChild, ElementRef } from '@angular/core';
import { PickerBaseComponent } from './pickerBase.component';
import { PickerDisplayMode } from 'filestack-js';
export class PickerOverlayComponent extends PickerBaseComponent {
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        const { filestackService, uploadSuccess, pickerOptions } = this;
        // Overwrite display mode to be always 'overlay' in this component
        this.picker = filestackService.picker(Object.assign({}, pickerOptions, { displayMode: PickerDisplayMode.overlay, onUploadDone: (/**
             * @param {?} res
             * @return {?}
             */
            res => uploadSuccess.next(res)) }));
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        const { picker, customRenderRef, uploadError } = this;
        if (!customRenderRef.nativeElement.children.length) {
            // Picker open success handler there is ommited, because it's accessible from pickerOptions
            picker.open()
                .then(void (0))
                .catch((/**
             * @param {?} err
             * @return {?}
             */
            err => uploadError.next(err)));
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        const { picker, uploadError } = this;
        event.stopPropagation();
        event.preventDefault();
        // Picker open success handler there is ommited, because it's accessible from pickerOptions
        picker.open()
            .then(void (0))
            .catch((/**
         * @param {?} err
         * @return {?}
         */
        err => uploadError.next(err)));
    }
}
PickerOverlayComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-picker-overlay',
                template: '<div #customRender><ng-content class="ng-picker"></ng-content></div>',
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PickerOverlayComponent.propDecorators = {
    customRenderRef: [{ type: ViewChild, args: ['customRender', { static: false },] }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    PickerOverlayComponent.prototype.customRenderRef;
}
